core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DetermineJoinDistributionType.java

```diff
@@ -105,20 +111,92 @@ public static double getSourceTablesSizeInBytes(PlanNode node, Context context)
 
     @VisibleForTesting
     static double getSourceTablesSizeInBytes(PlanNode node, Lookup lookup, StatsProvider statsProvider, TypeProvider typeProvider)
+    {
+        List<PlanNode> sourceNodes = getSourceNodes(node, lookup);
+        if (sourceNodes.isEmpty()) {
+            return NaN;
+        }
+
+        return sourceNodes.stream()
+                .mapToDouble(sourceNode -> statsProvider.getStats(sourceNode).getOutputSizeInBytes(sourceNode.getOutputSymbols(), typeProvider))
+                .sum();
+    }
+
+    private static double getApproximateSourceSizeInBytes(PlanNode node, Context context)
+    {
+        return getApproximateSourceSizeInBytes(node, context.getLookup(), context.getStatsProvider(), context.getSymbolAllocator().getTypes());
+    }
+
+    @VisibleForTesting
+    static double getApproximateSourceSizeInBytes(PlanNode node, Lookup lookup, StatsProvider statsProvider, TypeProvider typeProvider)
+    {
+        List<PlanNode> sourceNodes = getSourceNodes(node, lookup);
+        if (sourceNodes.isEmpty()) {
+            return NaN;
+        }
+
+        double sourceTablesSizeInBytes = sourceNodes.stream()
+                .mapToDouble(sourceNode -> statsProvider.getStats(sourceNode).getOutputSizeInBytes(sourceNode.getOutputSymbols(), typeProvider))
+                .sum();
+        if (isNaN(sourceTablesSizeInBytes)) {
+            return NaN;
+        }
+
+        double sourceTablesRowCount = sourceNodes.stream()
+                .mapToDouble(sourceNode -> statsProvider.getStats(sourceNode).getOutputRowCount())
+                .sum();
+
+        double firstKnownRowCount = getFirstKnownRowCount(node, lookup, statsProvider);
```

Comment:
I would just take `first know output size`?

> I was concerned about underestimation in cases where the symbol stats might be missing for some columns (e.g. due to unestimated terms like aggregates in AggregationStatsRule).

I don't fully understand this. Missing symbols stats should make size estimation `NaN`, right?

Reply:
Looking at `getOutputSizeForSymbol`, it seems we will always give some output size if outputRowCount is not `NaN`, so I've simplified the code to use `getOutputSizeInBytes` now.